### PR TITLE
Perform HEAD request for HttpStore::head

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #.

# Rationale for this change
 

Issuing a head request instead of a GET request allows HttpStore::head to be used against regular HTTP servers that don't speak WebDav.

# What changes are included in this PR?

# Are there any user-facing changes?

Technically perhaps, but we already enforced that GET requests return the necessary headers for metadata, and so this is just an evolution of that to HEAD requests.

### CODE REVIEW

**Comment by kylebarron** [comment]
A possible user-facing change for this is with signed urls, because those are only valid for either GET or HEAD (but not both) at one time, IIRC. So sometimes using a `GET` request instead of `HEAD` is nicer. Related to https://github.com/kylebarron/parquet-wasm/issues/272

**Comment by tustvold** [comment]
TBC previously this issued a WebDav specific PROPFIND request not a GET request

**Comment by universalmind303** [comment]
LGTM, We ([GlareDB](https://github.com/GlareDB)) are already running a wrapper around objectstore for this exact reason. Using HEAD seems much more flexible than PROPFIND.

File: object_store/src/client/get.rs [hunk 1]
```diff
@@ -49,10 +55,12 @@ impl<T: GetClient> GetClientExt for T {
     async fn get_opts(&self, location: &Path, options: GetOptions) -> Result<GetResult> {
         let range = options.range.clone();
         let response = self.get_request(location, options, false).await?;
-        let meta = header_meta(location, response.headers(), Default::default())
-            .map_err(|e| Error::Generic {
-                store: T::STORE,
-                source: Box::new(e),
+        let meta =
+            header_meta(location, response.headers(), T::HEADER_CONFIG).map_err(|e| {
```
Comment (alamb) [inline object_store/src/client/get.rs:59]: It took me a while to track down that the default for `HEADER_CONFIG` is the same as `HeaderConfig::default` and thus this isn't a change for `S3Client` and other implementations

File: object_store/src/client/get.rs [hunk 2]
```diff
@@ -28,6 +28,12 @@ use reqwest::Response;
 pub trait GetClient: Send + Sync + 'static {
     const STORE: &'static str;
 
+    /// Configure the [`HeaderConfig`] for this client
+    const HEADER_CONFIG: HeaderConfig = HeaderConfig {
+        etag_required: true,
+        last_modified_required: true,
+    };
```
Comment (alamb) [inline object_store/src/client/get.rs:35]: I think using `Default` would make the intent clearer here

```suggestion
    /// Configure the [`HeaderConfig`] used for requests issued by this client
    const HEADER_CONFIG: HeaderConfig = HeaderConfig::default();
```

File: object_store/src/http/client.rs [hunk 3]
```diff
@@ -307,6 +277,58 @@ impl Client {
     }
 }
 
+#[async_trait]
+impl GetClient for Client {
+    const STORE: &'static str = "HTTP";
+
+    const HEADER_CONFIG: HeaderConfig = HeaderConfig {
```
Comment (alamb) [inline object_store/src/http/client.rs:284]: Perhaps you can add an explanatory comment about why this client overrides the default HEADER_CONFIG (so as to support http servers that don't implement WEBDAV?)

**Review by alamb — APPROVED** [review]
I reviewed the code carefully and I think this change makes sense to me. 

Is there any way to test this change (with the standard admonition "how would we know if we broke the behavior in a future refactoring")?

File: object_store/src/client/get.rs (hunk 2 above)
Comment (tustvold) [reply object_store/src/client/get.rs:35]: Sadly this isn't possible as Default is not const

**Pull request merged**